src/Compilers/Core/Portable/PEWriter/PeWriter.cs

```diff
@@ -88,10 +89,17 @@ internal sealed class PeWriter
             // If PDB writer is given, we have to have PDB path.
             Debug.Assert(nativePdbWriterOpt == null || pdbPathOpt != null);
 
-            var peWriter = new PeWriter(context.Module.Properties, context.Module.Win32Resources, context.Module.Win32ResourceSection, pdbPathOpt, deterministic);
-            var mdWriter = FullMetadataWriter.Create(context, messageProvider, allowMissingMethodBodies, deterministic, getPortablePdbStreamOpt != null, cancellationToken);
+            try
+            {
+                var peWriter = new PeWriter(context.Module.Properties, context.Module.Win32Resources, context.Module.Win32ResourceSection, pdbPathOpt, deterministic);
+                var mdWriter = FullMetadataWriter.Create(context, messageProvider, allowMissingMethodBodies, deterministic, getPortablePdbStreamOpt != null, cancellationToken);
 
-            return peWriter.WritePeToStream(mdWriter, getPeStream, getPortablePdbStreamOpt, nativePdbWriterOpt);
+                return peWriter.WritePeToStream(mdWriter, getPeStream, getPortablePdbStreamOpt, nativePdbWriterOpt);
+            }
+            catch (Exception ex) when (!(ex is PdbWritingException || ex is ResourceException || ex is PermissionSetFileReadException))
```

Comment:
I agree that the hole should be closed. There are other ways to do it though. As I explained to @TyOverby I'd prefer the try catch to be in CommonCompiler around the call to Compilation.Emit. There we should add a filter that reports a non-fatal Watson and converts all exceptions to diagnostics. That would take care of the Watsons. 

I don't see the blanket catches. For example, ArgumentOutOfRangeException thrown by the metadata writer won't currently be caught anywhere (afaict). 

Also, the compiler is potentially invoking user code in PE writer (virtual methods on a Stream). We should let exceptions thrown by user-code caused by incorrect implementation pass thru, so that customer that hosts Roslyn in an app can gather crash dumps with actionable information for their own code. Unless we document otherwise for specific kinds of exceptions, as we do in MetadataReference.GetMetadata().